### ISSUE TITLE
Fix WatsonGateway from always refreshing the token

### ISSUE THREAD

**Issue by glennrfisher** [watson-developer-cloud/ios-sdk#130]
When multiple requests are in-flight, only the first request to return a 401 error should prompt the `authStrategy` to refresh its token. All others should be cached and properly reference the `authStrategy` so that they use the new token when restarted.

Right now, there is a bug that causes the token to be refreshed for every network request sent through `WatsonGateway`. (All operations are successful, but make three network calls when one is usually sufficient.)

The problem is probably caused by a pass-by-value call that prevents requests from pointing to the same shared `authStrategy`.

**Comment by glennrfisher** [comment]
After e7ab8eaa2e1f054aec0c9a3cf0c13ebab75050af I inserted print statements into `Dialog`, `WatsonGateway`, and `BasicAuthenticationStrategy` to print the address of the `authStrategy` along with the caller. The results are below.

The results of the test show that this problem has been resolved.

(Note that tokens are forcibly refreshed for upload/download requests that still use Alamofire.)

```
Test Case '-[WatsonDeveloperCloudTests.DialogTests testDialog]' started.
0x00007ff3fb81f900 - getDialogs
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - createDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogs
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogFile - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogFile - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogFile - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogFile - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - converse
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - converse
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateContent
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - createDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - createDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - createDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - deleteDialog
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getDialogFile - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateDialog - forced refresh
0x00007ff3fb81f900 - refreshToken
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - converse
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - converse
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - converse
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - getProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - updateProfile
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
0x00007ff3fb81f900 - deleteDialog
0x00007ff3fb81f900 - request
0x00007ff3fb81f900 - URLRequest
Test Case '-[WatsonDeveloperCloudTests.DialogTests testDialog]' passed (75.054 seconds).
```

**Comment by glennrfisher** [comment]
Performing the same experiment with `PersonalityInsights` led to a very interesting conclusion about Xcode tests.

As you can see from the results below, each test case has a different memory address for its associated `AuthenticationStrategy`. This means a new `AuthenticationStrategy` — and more importantly, a new `PersonalityInsights` service — was instantiated for each test.

The conclusion: *Each test in Xcode is completely independent. That is, there is no shared state between any tests.*

This invalidates the assumptions of #132, where we assumed that class-level properties were shared among tests.

```
Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testContentItem]' started.
0x00007ff152448f80 - PersonalityInsights init
0x00007ff152448f80 - getProfileContentItems
0x00007ff152448f80 - request
0x00007ff152448f80 - refreshToken
0x00007ff152448f80 - request
0x00007ff152448f80 - URLRequest
Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testContentItem]' passed (1.200 seconds).

Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testProfile]' started.
0x00007ff152445520 - PersonalityInsights init
0x00007ff152445520 - getProfileText
0x00007ff152445520 - request
0x00007ff152445520 - refreshToken
0x00007ff152445520 - request
0x00007ff152445520 - URLRequest
Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testProfile]' passed (0.766 seconds).

Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testProfileWithShortText]' started.
0x00007ff15271cde0 - PersonalityInsights init
0x00007ff15271cde0 - getProfileText
0x00007ff15271cde0 - request
0x00007ff15271cde0 - refreshToken
0x00007ff15271cde0 - request
0x00007ff15271cde0 - URLRequest
Test Case '-[WatsonDeveloperCloudTests.PersonalityInsightsTests testProfileWithShortText]' passed (0.441 seconds).
```

**Comment by glennrfisher** [comment]
Before closing this ticket we should investigate how `WatsonGateway` performs when the `AuthenticationStrategy` is implemented as a struct (passed-by-value) instead of a class (passed-by-reference).

I believe the struct would prevent shared state, meaning a request that refreshes the token cannot share that token with other requests. In the case of an expired token, this would result in *every* request having to refresh the token.

If that's the case, then we will have to update the code to always pass `AuthenticationStrategy` arguments by reference using `inout`.

**Comment by glennrfisher** [comment]
### The Problem

As expected, implementing an `AuthenticationStrategy` as a struct prevents shared state. Without shared state, the `WatsonGateway` enters an infinite loop when given a nil or expired token.

(`WatsonGateway` tries to cache the request, refresh the token, and increment the number of `retries`. With no shared state, however, not only does the cached request not receive the updated token, but its `retries` remains at zero. Since `retries < maxRetries` for the cached request, `WatsonGateway` will try to refresh the token again when the request is restarted, ad infinitum.)

### The Solution

To fix this, we should always declare the `authStrategy` argument as an `inout` to ensure that it is passed by reference, even when the `authStrategy` is implemented as a struct.

### The Experiment

You can follow these steps if you'd like to recreate the issue and confirm my results:

1. Retrieve an authentication token for the Dialog service.
2. Update DialogTests's `setup` function to create an `APIKeyAuthenticationStrategy` using the token you obtained. Pass this `APIKeyAuthenticationStrategy` object into the Dialog init instead of providing a username and password.
3. Remove all forced token refreshes within the Dialog service (i.e. in `createDialog`, `getDialogFile`, and `updateDialog`).
4. Remove `self.token = apiKey` from `APIKeyAuthenticationStrategy` so that the `token` property will be nil upon initialization. This will force `WatsonGateway` to execute the `refreshToken` function.
5. Update `WatsonGateway` to print "refreshing nil token" whenever the `authStrategy`'s token is nil. (i.e. within the `guard authStrategy.token != nil else { ... }` block).

Now you can switch `APIKeyAuthenticationStrategy` between a class or struct, run the tests for the Dialog service, and see how the behavior differs.

#### `APIKeyAuthenticationStrategy` as a Class

You will find that the Dialog tests execute as expected when `APIKeyAuthenticationStrategy` is implemented as a class: the token is initially nil, it is refreshed once, then shared among all endpoints/operations/cached requests, resulting in successful completion of the tests.

```
Test Case '-[WatsonDeveloperCloudTests.DialogTests testDialog]' started.
refreshing nil token
Test Case '-[WatsonDeveloperCloudTests.DialogTests testDialog]' passed (43.588 seconds).
```

#### `APIKeyAuthenticationStrategy` as a Struct

When `APIKeyAuthenticationStrategy` is instead implemented as a struct: the token is initially nil, it is refreshed, but the refreshed token value is not shared among all endpoints/operations/cached requests, causing an infinite loop and the tests to fail.

```
Test Case '-[WatsonDeveloperCloudTests.DialogTests testDialog]' started.
refreshing nil token
refreshing nil token
refreshing nil token
refreshing nil token
refreshing nil token
refreshing nil token
...
refreshing nil token
warning: could not load any Objective-C class information. This will significantly reduce the quality of type information available.
(lldb)
```

**Comment by glennrfisher** [comment]
I will update all `authStrategy` arguments to be `inout` then re-run the experiment above to verify the fix.

**Comment by glennrfisher** [comment]
Even after I changed all `authStrategy` arguments to `inout`, I saw the same infinite-loop behavior with struct implementations of `AuthenticationStrategy`. Not sure why (perhaps it's a bug on my part or because the struct — which is allocated on the stack — falls out of scope).

Regardless, I found [Apple's recommendation](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Protocols.html#//apple_ref/doc/uid/TP40014097-CH25-ID281) for our problem: "Use a class-only protocol when the behavior defined by that protocol’s requirements assumes or requires that a conforming type has reference semantics rather than value semantics."

So to resolve this issue, I will update the `AuthenticationStrategy` protocol to be class-only.
